crypto/ed25519/ed25519.go

```diff
@@ -170,3 +171,31 @@ func (pubKey PubKey) Equals(other crypto.PubKey) bool {
 
 	return false
 }
+
+var _ crypto.BatchVerifier = BatchVerifier{}
+
+type BatchVerifier struct {
+	ed25519consensus.BatchVerifier
+}
+
+func NewBatchVerifier() crypto.BatchVerifier {
+	return BatchVerifier{ed25519consensus.NewBatchVerifier()}
+}
+
+func (b BatchVerifier) Add(key crypto.PubKey, msg, signature []byte) error {
+	if l := len(key.Bytes()); l != PubKeySize {
+		return errors.New("pubkey size is incorrect")
+	}
+
+	if len(signature) != SignatureSize || signature[63]&224 != 0 {
+		return errors.New("invalid signature")
+	}
+
+	b.BatchVerifier.Add(ed25519.PublicKey(key.Bytes()), msg, signature)
```

Comment:
why do we copy signature below, but does not do the same here?

Reply:
I dont see a copy below. Do you mean on pubkey?

Reply:
var sig64 [SignatureSize]byte
	copy(sig64[:], sig)

in another file

Reply:
oh in sr25519 they are asking for a sized byte array while here it asks for an unsized one. Within the ed25519 implementation since we split the signature into r and s values (r = sig[:32], s = sig[32:]) we don't need to size it ourselves.